src/device_trezor/trezor/protocol.cpp

```diff
@@ -0,0 +1,896 @@
+    }
+  }
+
+  static void generate_rsig_batch_sizes(std::vector<uint64_t> &batches, unsigned rsig_type, size_t num_outputs){
+    size_t amount_batched = 0;
+
+    while(amount_batched < num_outputs){
+      if (rsig_type == rct::RangeProofBorromean || rsig_type == rct::RangeProofBulletproof) {
+        batches.push_back(1);
+        amount_batched += 1;
+
+      } else if (rsig_type == rct::RangeProofPaddedBulletproof){
+        if (num_outputs > BULLETPROOF_MAX_OUTPUTS){
+          throw std::invalid_argument("BP padded can support only BULLETPROOF_MAX_OUTPUTS statements");
+        }
+        batches.push_back(num_outputs);
+        amount_batched += num_outputs;
+
+      } else if (rsig_type == rct::RangeProofMultiOutputBulletproof){
+        size_t batch_size = 1;
+        while (batch_size * 2 + amount_batched <= num_outputs && batch_size * 2 <= BULLETPROOF_MAX_OUTPUTS){
+          batch_size *= 2;
+        }
+        batch_size = std::min(batch_size, num_outputs - amount_batched);
+        batches.push_back(batch_size);
+        amount_batched += batch_size;
+
+      } else {
+        throw std::invalid_argument("Unknown rsig type");
+      }
+    }
+  }
+
+  void Signer::compute_integrated_indices(TsxData * tsx_data){
+    if (m_aux_data == nullptr || m_aux_data->tx_recipients.empty()){
+      return;
+    }
+
+    auto & chg = tsx_data->change_dts();
+    std::string change_hash = hash_addr(&chg.addr(), chg.amount(), chg.is_subaddress());
+
+    std::vector<uint32_t> integrated_indices;
+    std::set<std::string> integrated_hashes;
+    for (auto & cur : m_aux_data->tx_recipients){
+      if (!cur.has_payment_id){
+        continue;
+      }
+      integrated_hashes.emplace(hash_addr(&cur.address.m_spend_public_key, &cur.address.m_view_public_key));
+    }
+
+    ssize_t idx = -1;
+    for (auto & cur : tsx_data->outputs()){
+      idx += 1;
+
+      std::string c_hash = hash_addr(&cur.addr(), cur.amount(), cur.is_subaddress());
+      if (c_hash == change_hash || cur.is_subaddress()){
+        continue;
+      }
+
+      c_hash = hash_addr(&cur.addr());
+      if (integrated_hashes.find(c_hash) != integrated_hashes.end()){
+        integrated_indices.push_back((uint32_t)idx);
+      }
+    }
+
+    if (!integrated_indices.empty()){
+      assign_to_repeatable(tsx_data->mutable_integrated_indices(), integrated_indices.begin(), integrated_indices.end());
+    }
+  }
+
+  std::shared_ptr<messages::monero::MoneroTransactionInitRequest> Signer::step_init(){
+    // extract payment ID from construction data
+    auto & tsx_data = m_ct.tsx_data;
+    auto & tx = cur_tx();
+
+    m_ct.tx.version = 2;
+    m_ct.tx.unlock_time = tx.unlock_time;
+
+    tsx_data.set_version(1);
+    tsx_data.set_unlock_time(tx.unlock_time);
+    tsx_data.set_num_inputs(static_cast<google::protobuf::uint32>(tx.sources.size()));
+    tsx_data.set_mixin(static_cast<google::protobuf::uint32>(tx.sources[0].outputs.size()));
```

Comment:
This seems like it might be wrong: mixin is the number of fake outs, so one less than the ring size, and outputs likely contains all the ring members.